packages/unmock-node/src/backend/index.ts

```diff
@@ -24,58 +34,148 @@ async function handleRequestAndResponse(
   createResponse: CreateResponse,
   req: IncomingMessage,
   res: ServerResponse,
+  clientRequest: ClientRequest, // For emitting errors
 ) {
   try {
     const serializedRequest: ISerializedRequest = await serializeRequest(req);
+    debugLog("Serialized request", JSON.stringify(serializedRequest));
     const serializedResponse: ISerializedResponse | undefined = createResponse(
       serializedRequest,
     );
+
     if (serializedResponse === undefined) {
-      throw Error(constants.MESSAGE_FOR_MISSING_MOCK);
+      // TODO Handle this properly
+      debugLog("No match found, emitting error");
+      clientRequest.emit("error", Error("No matching template found"));
+      return;
     }
     respondFromSerializedResponse(serializedResponse, res);
   } catch (err) {
-    // TODO Emit an error in the corresponding client request instead?
-    const errorResponse: ISerializedResponse = {
-      body: err.message,
-      statusCode: constants.STATUS_CODE_FOR_ERROR,
-    };
-    respondFromSerializedResponse(errorResponse, res);
+    clientRequest.emit("error", Error(`unmock error: ${err.message}`));
   }
 }
 
 export interface INodeBackendOptions {
   servicesDirectory?: string;
 }
 
+interface IBypassableSocket extends net.Socket {
+  bypass: () => void;
+}
+
 const nodeBackendDefaultOptions: INodeBackendOptions = {};
+const UNMOCK_INTERNAL_HTTP_HEADER = "x-unmock-req-id";
 
-let mitm: any;
 export default class NodeBackend implements IBackend {
   private readonly config: INodeBackendOptions;
+  private clientRequests: {
+    [requestId: string]: {
+      clientRequest: ClientRequest;
+    };
+  } = {};
+  private origOnSocket?: (socket: net.Socket) => void;
+  private mitm: any;
+
   public constructor(config?: INodeBackendOptions) {
     this.config = { ...nodeBackendDefaultOptions, ...config };
   }
 
   public initialize(options: UnmockOptions) {
-    mitm = Mitm();
-    mitm.on("connect", (socket: any, opts: any) => {
-      if (options.isWhitelisted(opts.host)) {
-        socket.bypass();
-      }
-    });
+    if (this.mitm !== undefined) {
+      this.reset();
+    }
+    this.mitm = Mitm();
+
+    const self = this;
+    this.origOnSocket = ClientRequest.prototype.onSocket;
+
+    /**
+     * When a socket is assigned to a client request, create an internal ID
+     * and add the ID in the request header. Thereby we can map
+     * the server-side incoming request to the corresponding
+     * client request and emit errors properly.
+     * Borrowed from https://github.com/FormidableLabs/yesno.
+     */
+    ClientRequest.prototype.onSocket = _.flowRight(
```

Comment:
Just as a side note, as there are only two functions here, I'm sure we could get away without requiring `lodash` to pass value from one function to the other... :)
Is this how `yesno` does it as well?

Reply:
Yes that's how yesno does it, it's just a pretty way to do chaining.